File: backend/app/api/workflows.py
"""Workflow 能力层接口（Step 4）。

能力矩阵是这一步的核心视图：四种能力各自「就绪 / 缺什么 / 缺了会影响什么」，
以及 ComfyUI 在不在线。校验分本地绑定检查与节点探测两段，离线也能用前一半。
"""

from __future__ import annotations

from typing import Any, Literal

from fastapi import APIRouter
from pydantic import BaseModel, Field

from app.services.workflows import workflows

router = APIRouter(tags=["workflows"])


class ImportBody(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    capability: str = Field(description="text2image / image2video / first_last_frame / upscale")
    api_json: str = Field(description="ComfyUI 导出的 workflow_api.json 原文")
    bindings: dict[str, Any] | None = Field(
        default=None, description='槽位 → "节点id.字段名"，例如 {"prompt": "6.text"}'
    )
    notes: str | None = None


class BindBody(BaseModel):
    bindings: dict[str, Any]


class ProjectBindingsBody(BaseModel):
    generation_mode: Literal["comfy_preset", "http_api", "workflow_api"] = Field(
        default="comfy_preset", description="comfy_preset / http_api / workflow_api"
    )
    text2image: str | None = None
    image2video: str | None = None
    first_last_frame: str | None = None
    upscale: str | None = None


class UpdateBody(BaseModel):
    name: str | None = None
    notes: str | None = None
    capability: str | None = None
    status: str | None = Field(default=None, description="只允许改为 disabled / draft")


@router.get("/workflows")
async def list_global_workflows() -> list[dict[str, Any]]:
    return await workflows.list_global()


@router.get("/capabilities")
async def global_capability_matrix() -> dict[str, Any]:
    return await workflows.global_capability_matrix()


@router.get("/workflows/{wid}")
async def get_global_workflow(wid: str) -> dict[str, Any]:
    return await workflows.get_global(wid)


@router.post("/workflows", status_code=201)
async def import_global_workflow(body: ImportBody) -> dict[str, Any]:
    return await workflows.import_global(
        name=body.name,
        capability=body.capability,
        api_json=body.api_json,
        bindings=body.bindings,
        notes=body.notes,
    )


@router.put("/workflows/{wid}/bindings")
async def bind_global_workflow(wid: str, body: BindBody) -> dict[str, Any]:
    return await workflows.bind_global(wid, body.bindings)


@router.post("/workflows/{wid}/validate")
async def validate_global_workflow(wid: str, probe: bool = True) -> dict[str, Any]:
    return await workflows.validate_global(wid, probe=probe)


@router.post("/workflows/{wid}/default")
async def set_global_default(wid: str) -> dict[str, Any]:
    return await workflows.set_default_global(wid)


@router.patch("/workflows/{wid}")
async def update_global_workflow(wid: str, body: UpdateBody) -> dict[str, Any]:
    return await workflows.update("", wid, body.model_dump(exclude_none=True))


@router.delete("/workflows/{wid}", status_code=204)
async def delete_global_workflow(wid: str) -> None:
    await workflows.delete_global(wid)


@router.get("/projects/{pid}/workflow-bindings")
async def get_project_workflow_bindings(pid: str) -> dict[str, str | None]:
    return await workflows.project_bindings(pid)


@router.put("/projects/{pid}/workflow-bindings")
async def set_project_workflow_bindings(
    pid: str, body: ProjectBindingsBody
) -> dict[str, str | None]:
    return await workflows.set_project_bindings(pid, body.model_dump())


@router.get("/projects/{pid}/workflows")
async def list_workflows(pid: str) -> list[dict[str, Any]]:
    return await workflows.list_workflows(pid)


@router.get("/projects/{pid}/capabilities")
async def capability_matrix(pid: str) -> dict[str, Any]:
    return await workflows.project_capabilities(pid)


@router.post("/projects/{pid}/workflows", status_code=201)
async def import_workflow(pid: str, body: ImportBody) -> dict[str, Any]:
    return await workflows.import_workflow(
        pid,
        name=body.name,
        capability=body.capability,
        api_json=body.api_json,
        bindings=body.bindings,
        notes=body.notes,
    )


@router.get("/projects/{pid}/workflows/{wid}")
async def get_workflow(pid: str, wid: str) -> dict[str, Any]:
    return await workflows.get(pid, wid)


@router.patch("/projects/{pid}/workflows/{wid}")
async def update_workflow(pid: str, wid: str, body: UpdateBody) -> dict[str, Any]:
    return await workflows.update(pid, wid, body.model_dump(exclude_none=True))


@router.put("/projects/{pid}/workflows/{wid}/bindings")
async def bind_workflow(pid: str, wid: str, body: BindBody) -> dict[str, Any]:
    return await workflows.bind(pid, wid, body.bindings)


@router.post("/projects/{pid}/workflows/{wid}/validate")
async def validate_workflow(pid: str, wid: str, probe: bool = True) -> dict[str, Any]:
    return await workflows.validate(pid, wid, probe=probe)


@router.post("/projects/{pid}/workflows/{wid}/default")
async def set_default(pid: str, wid: str) -> dict[str, Any]:
    return await workflows.set_default(pid, wid)


@router.delete("/projects/{pid}/workflows/{wid}", status_code=204)
async def delete_workflow(pid: str, wid: str) -> None:
    await workflows.delete(pid, wid)

File: frontend/src/shared/api/workflows.ts
/**
 * Workflow 能力层接口（Step 4）。
 *
 * 这一层是「业务不绑定具体视频模型」那条硬约束的落点：镜头只说要什么能力
 * （`text2image` / `image2video` / `first_last_frame` / `upscale`），
 * 哪套图、哪个节点的哪个字段接 prompt，全在这里的绑定表里。
 *
 * 两个形状上的要点：
 *   1. **校验分两段**——`validate?probe=false` 只做本地绑定检查（离线可用），
 *      `probe=true` 才去 ComfyUI 探测自定义节点。所以 ComfyUI 不在线也能把绑定配对。
 *   2. **校验不通过是结构化错误**——`INVALID_WORKFLOW` / `COMFY_NODE_MISSING`
 *      会带 `problems` / `missing_nodes`，页面照常显示 suggestions，不要吞掉。
 */

import { api } from './client'

/** 四种能力，顺序与后端 `models_gen.CAPABILITIES` 一致。 */
export const CAPABILITIES = ['text2image', 'image2video', 'first_last_frame', 'upscale'] as const
export type Capability = (typeof CAPABILITIES)[number]

export const CAPABILITY_LABEL: Record<Capability, string> = {
  text2image: '文生图',
  image2video: '图生视频',
  first_last_frame: '首尾帧',
  upscale: '超分',
}

/** 可绑定的槽位。哪些是必填由能力决定（`capability_matrix` 里的 `required_slots`）。 */
export const SLOTS = [
  'prompt',
  'negative_prompt',
  'reference_image',
  'first_frame',
  'last_frame',
  'source_image',
  'seed',
  'steps',
  'width',
  'height',
  'duration',
] as const
export type Slot = (typeof SLOTS)[number]

/** 图里的一个节点。`fields` 只含标量字段——连线出来的输入不能被绑定覆盖。 */
export interface WorkflowNode {
  id: string
  class_type: string
  title: string | null
  fields: string[]
}

/** 一次校验的结果。也会被后端写进 `validation_json`，所以列表里就能看到上次结果。 */
export interface ValidationResult {
  ok: boolean
  problems: string[]
  missing_slots: string[]
  required_nodes: string[]
  missing_nodes: string[]
  /** 「已探测 N 个节点」/「已跳过节点探测」/「未能探测（…）」 */
  probe: string
  checked_at: string
}

export interface Workflow {
  id: string
  name: string
  capability: string
  /** draft / ready / invalid / disabled —— 只有 ready 的能被 `resolve` 选中。 */
  status: string
  is_default: number
  notes: string | null
  /** 槽位 → `"节点id.字段名"`。 */
  bindings: Record<string, string>
  reference_image_slots?: string[]
  reference_image_count?: number
  nodes: WorkflowNode[]
  required_nodes: string[]
  validation: ValidationResult | null
  missing_slots: string[]
  created_at: string
  updated_at: string
  /** 只有 `GET /workflows/{wid}` 带原始 API 图，列表里没有。 */
  api_json?: string
}

/** 能力矩阵的一行。`impact` 是「缺了会做不出什么」，没缺时为 null。 */
export interface CapabilityRow {
  capability: string
  ready: boolean
  workflow_count: number
  ready_count: number
  default_workflow_id: string | null
  default_workflow_name: string | null
  required_slots: string[]
  impact: string | null
}

export interface ComfyStatus {
  online: boolean
  base_url: string
  detail: string
}

export interface CapabilityMatrix {
  capabilities: CapabilityRow[]
  comfy: ComfyStatus
  project_bindings?: ProjectWorkflowBindings
}

export type GenerationMode = 'comfy_preset' | 'http_api' | 'workflow_api'
export interface ProjectWorkflowBindings {
  generation_mode: GenerationMode
  text2image: string | null
  image2video: string | null
  first_last_frame: string | null
  upscale: string | null
}

export interface ImportWorkflowBody {
  name: string
  capability: string
  api_json: string
  bindings?: Record<string, string>
  notes?: string | null
}

export const workflowsApi = {
  globalList: () => api.get<Workflow[]>('/workflows'),
  globalGet: (wid: string) => api.get<Workflow>(`/workflows/${wid}`),
  globalImport: (body: ImportWorkflowBody) => api.post<Workflow>('/workflows', body),
  globalBind: (wid: string, bindings: Record<string, string>) =>
    api.put<Workflow>(`/workflows/${wid}/bindings`, { bindings }),
  globalValidate: (wid: string, probe = true) =>
    api.post<ValidationResult>(`/workflows/${wid}/validate?probe=${probe ? 'true' : 'false'}`, {}),
  globalSetDefault: (wid: string) => api.post<Workflow>(`/workflows/${wid}/default`, {}),
  globalRemove: (wid: string) => api.del<void>(`/workflows/${wid}`),
  list: (_pid: string) => api.get<Workflow[]>('/workflows'),
  get: (_pid: string, wid: string) => api.get<Workflow>(`/workflows/${wid}`),
  matrix: (pid: string) =>
    api.get<CapabilityMatrix>(pid ? `/projects/${pid}/capabilities` : '/capabilities'),
  import: (_pid: string, body: ImportWorkflowBody) =>
    api.post<Workflow>('/workflows', body),
  update: (_pid: string, wid: string, patch: { name?: string; notes?: string; status?: string }) =>
    api.patch<Workflow>(`/workflows/${wid}`, patch),
  bind: (_pid: string, wid: string, bindings: Record<string, string>) =>
    api.put<Workflow>(`/workflows/${wid}/bindings`, { bindings }),
  /** `probe=false` 时不碰 ComfyUI，只查绑定；离线也能把工作流配到「就绪」以外的所有问题查清。 */
  validate: (_pid: string, wid: string, probe = true) =>
    api.post<ValidationResult>(
      `/workflows/${wid}/validate?probe=${probe ? 'true' : 'false'}`,
      {},
    ),
  setDefault: (_pid: string, wid: string) =>
    api.post<Workflow>(`/workflows/${wid}/default`, {}),
  remove: (_pid: string, wid: string) => api.del<void>(`/workflows/${wid}`),
  projectBindings: (pid: string) =>
    api.get<ProjectWorkflowBindings>(`/projects/${pid}/workflow-bindings`),
  setProjectBindings: (pid: string, bindings: ProjectWorkflowBindings) =>
    api.put<ProjectWorkflowBindings>(`/projects/${pid}/workflow-bindings`, bindings),
}
